scripts/generate_errors.pl

```diff
@@ -53,26 +53,54 @@
                              PEM PK PKCS12 PKCS5
                              RSA SSL X509 );
 
-my $line_separator = $/;
 undef $/;
 
 open(FORMAT_FILE, '<:crlf', "$error_format_file") or die "Opening error format file '$error_format_file': $!";
 my $error_format = <FORMAT_FILE>;
 close(FORMAT_FILE);
 
-$/ = $line_separator;
-
 my @files = <$include_dir/*.h>;
 my @necessary_include_files;
 my @matches;
 foreach my $file (@files) {
     open(FILE, '<:crlf', "$file");
-    my @grep_res = grep(/^\s*#define\s+MBEDTLS_ERR_\w+\s+\-0x[0-9A-Fa-f]+/, <FILE>);
-    push(@matches, @grep_res);
+    my $content = <FILE>;
     close FILE;
-    my $include_name = $file;
-    $include_name =~ s!.*/!!;
-    push @necessary_include_files, $include_name if @grep_res;
+    my $found = 0;
+    while ($content =~ m[
+            # Both the before-comment and the after-comment are optional.
+            # Only the comment content is a regex capture group. The comment
+            # start and end parts are outside the capture group.
+            (?:/\*[*!](?!<)             # Doxygen before-comment start
+                ((?:[^*]|\*+[^*/])*)    # $1: Comment content (no */ inside)
```

Comment:
I believe the same effect (here and for `$4`) could be achieved using a non-greedy operator: `(.*?)` which would IMO be more readable. Am I missing something?

Reply:
I made the same mistake when I was writing this code. I hesitated in writing a comment, because “why the code isn't written this other way” (which some readers would consider obviously inapplicable) comments can be a distraction sometimes. I guess I should have.

A non-greedy operator means that the match _ends_ as soon as it can. That works for the after-comment. That doesn't help with the before-comment which still obeys the rule that the whole regex starts matching at the first opportunity, so e.g. the first hit starts at the `/**` at the top of the file, and the non-greediness means that the `#define MBEDTLS_ERR` part is the first such definition, rather than the last definition if it was a greedy match.

Reply:
Thanks for the explanation. Ok, that's quite surprising at first, but then it becomes clear when you remember non-greedy means "end match at the first opportunity" (but still start it at the first opportunity) and not "find the shortest possible match" (which would be quite unreasonable when you think about it).